File: go/2025/01.go
package y2025

import (
	"bufio"
	"io"
	"strconv"

	"github.com/Grub4K/aoc-solutions/go/aoc"
)

func day01(r io.Reader) aoc.Result {
	scanner := bufio.NewScanner(r)
	scanner.Split(bufio.ScanLines)

	dial := 50
	counterA := 0
	counterB := 0
	clockwise := true
	for scanner.Scan() {
		line := scanner.Text()
		value, err := strconv.ParseUint(line[1:], 10, 32)
		if err != nil {
			return aoc.Error(err)
		}
		movement := int(value)

		if line[0] == 'R' != clockwise {
			dial = (100 - dial) % 100
			clockwise = !clockwise
		}

		total := dial + movement
		counterB += total / 100
		dial = total % 100
		if dial == 0 {
			counterA++
		}
	}
	if err := scanner.Err(); err != nil {
		return aoc.Error(err)
	}

	return aoc.Solution(counterA, counterB)
}

File: python/2025/01.py
from __future__ import annotations

import enum


class Direction(enum.Enum):
    Clockwise = enum.auto()
    Counterclockwise = enum.auto()


def process_line(line: str):
    direction = Direction.Clockwise if line[0] == "R" else Direction.Counterclockwise
    return direction, int(line[1:])


def run(data: list[tuple[Direction, int]]):
    dial = 50
    counter_a = 0
    counter_b = 0
    prev_direction = Direction.Clockwise

    for direction, movement in data:
        if direction != prev_direction:
            dial = (100 - dial) % 100
            prev_direction = direction

        rotations, dial = divmod(dial + movement, 100)
        counter_a += dial == 0
        counter_b += rotations


    yield counter_a
    yield counter_b
